### ISSUE TITLE
Setting properties on nested objects

### ISSUE THREAD

**Issue by mochacoder** [AriaFallah/mobx-store#34]
I feel like I'm just missing something obvious here but I'm having trouble setting a nested object property in a way that mobx-store will retain the history:

If I have the following store:

`const store = mobxstore({
  todos: {
    id123: { title: 'test todo', complete: false }
  }
});`

The only way I can successfully update the title is if I use something like the following:

`store('todos').get("id123").title = "test update todo title";`

But it looks like that bypasses the built in mechanisms mobx-store has for doing sets on objects as title is updated in the history and not included as a separate update.

Again seems like I'm missing something obvious but the examples don't show anything nested like this. Thanks for the help.

**Comment by AriaFallah** [comment]
Yeah so I should probably document this better, but this is probably the biggest flaw with the library right now.

If you look at this issue https://github.com/mobxjs/mobx/issues/214, MobX currently has no way to do deep observation, and the only solution I can think of would make this library horribly complex and bug prone.

This mean that if you make any changes deeper than one level into the store, there isn't a way to undo them. The only workaround as of now is something like this:

```js
const todos = store('todos')
todos.set('id123', { ...todos.get('id123'), 'title: test update todo title' })
```

Basically, you have to take a redux like approach, and make changes in an immutable manner. This way you're changing the top level `id123`, and it'll be registered in the history.

The MobX roadmap has a time traveling mechanism listed, but for now this is the best I can do with the current undo/redo implementation.

Is that helpful?

**Comment by mweststrate** [comment]
@ariafallah, did you take a look at the spy api? time travelling will be
based on that. Main downside of spy is that it is global, so you have to
mark the objects you are interested in somehow, but besides that I think it
can be used for redo / undo, as the set of possible mutations is pretty
limited and they have all the info needed for undo / redo.

Op za 9 jul. 2016 om 16:06 schreef Aria Fallah <notifications@github.com>:

**Comment by AriaFallah** [comment]
@mweststrate 

Really good idea actually...I can look into this for the time being as well.

**Comment by tcjustin** [comment]
Thanks for the quick response @AriaFallah that was helpful.  @mweststrate yeah the spy API might help me out as well, didn't know it existed.

**Comment by AriaFallah** [comment]
@mochacoder @mweststrate 

Probably has some quirks, but for the time being it works 😄 

https://github.com/AriaFallah/mobx-store/pull/35

**Comment by vojto** [comment]
So this still doesn't fire schedule, or am I doing something wrong?

Trying to update local storage when data changes.

**Comment by AriaFallah** [comment]
@vojto 

Could you post the code you're trying to schedule? There are some quirks with when your tasks run so it'd be helpful to have the code.

The scheduling is based on autorun, which I go into a bit of depth about here: https://github.com/mobxjs/mobx/issues/248#issuecomment-218927070. Basically it's not uncommon that in certain cases an autorun will never fire.
